op-batcher/batcher/channel_manager_memory_test.go

```diff
@@ -0,0 +1,185 @@
+				comp := compressor
+				algorithm := algo
+				batchType := batch
+
+				t.Run(testName, func(t *testing.T) {
+					runMemoryTest(t, batchType.batchType, comp.name, algorithm.algo, comp.setupFunc)
+				})
+			}
+		}
+	}
+}
+
+func runMemoryTest(t *testing.T, batchType uint, compressorType string, compressionAlgo derive.CompressionAlgo, setupCompressor func(*ChannelConfig, derive.CompressionAlgo)) {
+	log := testlog.Logger(t, log.LevelCrit)
+
+	// Create a channel manager with small frame size to force multiple channels
+	// Use smaller frame size for span batches since they're more efficient at packing data
+	frameSize := uint64(1000)
+	if batchType == derive.SpanBatchType {
+		frameSize = 500 // Smaller frame size for span batches to create more channels
+	}
+	cfg := channelManagerTestConfig(frameSize, batchType)
+	cfg.ChannelTimeout = 100 // Reasonable timeout
+	setupCompressor(&cfg, compressionAlgo)
+
+	// Use the existing default test rollup config to ensure chain IDs match
+	m := NewChannelManager(log, metrics.NoopMetrics, cfg, defaultTestRollupConfig)
+	m.Clear(eth.BlockID{})
+
+	// Measure initial memory
+	var initialMem runtime.MemStats
+	runtime.GC()
+	runtime.ReadMemStats(&initialMem)
+
+	// Create the first block (genesis)
+	var prevBlock *types.Block
+
+	// Add many blocks to create multiple channels
+	const numBlocks = 1000
+	for i := 0; i < numBlocks; i++ {
+		var block *types.Block
+
+		if i == 0 {
+			block = newMiniL2BlockWithChainID(0, defaultTestRollupConfig.L2ChainID)
+		} else {
+			// Create a block with proper parent hash and transaction content
+			block = newMiniL2BlockWithChainIDNumberParentAndL1Information(5, defaultTestRollupConfig.L2ChainID, big.NewInt(int64(i)), prevBlock.Hash(), 100, 0)
+		}
+
+		require.NoError(t, m.AddL2Block(block))
+		prevBlock = block
+
+		// Periodically process blocks and create channels
+		if i%100 == 0 {
+			// Ensure we have a channel
+			require.NoError(t, m.ensureChannelWithSpace(eth.BlockID{
+				Hash:   block.Hash(),
+				Number: block.NumberU64(),
+			}))
+
+			// Process blocks into channels
+			require.NoError(t, m.processBlocks())
+
+			// Try to get transaction data to fill channels
+			_, err := m.TxData(eth.BlockID{}, false)
+			// It's okay if there's no data ready (io.EOF)
+			if err != nil && err.Error() != "EOF" {
+				require.NoError(t, err)
+			}
+		}
+	}
+
+	// Final processing to ensure all blocks are processed
+	require.NoError(t, m.ensureChannelWithSpace(eth.BlockID{}))
+	require.NoError(t, m.processBlocks())
+
+	// Measure final memory
+	var finalMem runtime.MemStats
+	runtime.GC()
+	runtime.ReadMemStats(&finalMem)
+
+	// Calculate memory used by the channel manager
+	memUsed := finalMem.Alloc - initialMem.Alloc
+
+	// Assert that memory usage doesn't exceed 512MB
+	const maxMemoryMB = 512 * 1024 * 1024 // 512MB in bytes
+	require.Less(t, memUsed, uint64(maxMemoryMB),
+		"Channel manager used %d bytes (%.2f MB), exceeding 512 MB limit",
```

Comment:
This test will fail here (for sub sub test cases) without the optimization on this PR.